subsys/emul/espi/emul_espi_host.c

```diff
@@ -0,0 +1,234 @@
+	{ ESPI_VWIRE_SIGNAL_SMI,           0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_RST_CPU_INIT,  0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_HOST_RST_ACK,  0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_HOST_RST_WARN, 0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SUS_ACK,       0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_DNX_ACK,       0, ESPI_SLAVE_TO_MASTER },
+	{ ESPI_VWIRE_SIGNAL_SUS_WARN,      0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SUS_PWRDN_ACK, 0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_A,         0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_LAN,       0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_SLP_WLAN,      0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_HOST_C10,      0, ESPI_MASTER_TO_SLAVE },
+	{ ESPI_VWIRE_SIGNAL_DNX_WARN,      0, ESPI_MASTER_TO_SLAVE },
+};
+
+/** Run-time data used by the emulator */
+struct espi_host_emul_data {
+	/** eSPI emulator detail */
+	struct espi_emul emul;
+	/** eSPI controller device */
+	const struct device *espi;
+	/** Configuration information */
+	const struct espi_host_emul_cfg *cfg;
+	/** Virtual Wires states, for one slave only.
+	 *  With multi-slaves config, the states should be saved per slave */
+	struct vw_data vw_state[ARRAY_SIZE(vw_state_default)];
+};
+
+/** Static configuration for the emulator */
+struct espi_host_emul_cfg {
+	/** Label of the eSPI bus this emulator connects to */
+	const char *espi_label;
+	/** Label of the emulated AP*/
+	const char *label;
+	/** Pointer to run-time data */
+	struct espi_host_emul_data *data;
+	/* eSPI chip-select of the emulated device */
+	uint16_t chipsel;
+};
+
+static void emul_host_init_vw_state(struct espi_host_emul_data *data)
+{
+	unsigned i;
+
+	for (i = 0; i < ARRAY_SIZE(data->vw_state); i++) {
+		data->vw_state[i] = vw_state_default[i];
+	}
+	return;
+}
+
+static int emul_host_find_index(struct espi_host_emul_data *data,
+				enum espi_vwire_signal vw)
+{
+	int idx;
+
+	for (idx = 0; idx < ARRAY_SIZE(data->vw_state); idx++) {
+		if (data->vw_state[idx].sig == vw) {
+			return idx;
+		}
+	}
+
+	return -1;
+}
+
+static int emul_host_set_vw(struct espi_emul *emul, enum espi_vwire_signal vw,
+			    uint8_t level)
+{
+	struct espi_host_emul_data *data;
+	int idx;
+
+	data = CONTAINER_OF(emul, struct espi_host_emul_data, emul);
+	idx = emul_host_find_index(data, vw);
+
+	if (idx < 0 || data->vw_state[idx].dir != ESPI_SLAVE_TO_MASTER) {
+		LOG_ERR("%s: invalid vw: %d", __func__, vw);
+		return -EIO;
+	}
+
+	data->vw_state[idx].level = level;
+
+	return 0;
+}
+
+static int emul_host_get_vw(struct espi_emul *emul, enum espi_vwire_signal vw,
+			    uint8_t *level)
+{
+	struct espi_host_emul_data *data;
+	int idx;
+
+	data = CONTAINER_OF(emul, struct espi_host_emul_data, emul);
+	idx = emul_host_find_index(data, vw);
+
+	if (idx < 0 || data->vw_state[idx].dir != ESPI_MASTER_TO_SLAVE) {
+		LOG_ERR("%s: invalid vw: %d", __func__, vw);
+		return -EIO;
```

Comment:
I wonder if EPERM might be better, or something else distinctive? With -EIO I think of an error on a bus. Perhaps that is what this is, but it seems to me more like a programming error.

Reply:
I agree, it is rather the programming error.